docs/testing-strategy.md

```diff
@@ -0,0 +1,154 @@
+
+As per common Rails conventions, most of the directories within the spec
+directory contain Ruby [unit tests](#unit-tests) with the directories matching
+those within the app directory. Code that is stored in the lib directory has
+corresponding tests in the spec/lib directory.
+
+Directories within spec that don't contain Ruby unit tests are as follows:
+
+- factories - contains [FactoryBot factories][factory-bot], which are a fixture
+  alternative;
+- fixtures - contains supporting files for tests (for example image files) which
+  aren't easily produced by factories;
+- features - contains [feature tests](#feature-tests), which exercises
+  application features from a user perspective;
+- javascripts - contains [unit tests](#unit-tests) for JavaScript files
+  and configuration and helper files for Jasmine;
+- requests - contains [request tests](#request-tests), which test the HTTP API
+  of the application;
+- support - contains helper files for shared test methods and logic;
+- views - contains [view tests](#view-tests), which are used to test
+  [Rails views][rails-views] output expected HTML in a given scenario.
+
+## Unit tests
+
+The purpose of these tests is to test individual units of the system in
+isolation. They are intended to provide exhaustive tests of the code paths
+through a particular class or module through its public interface. As per
+the [test pyramid](https://martinfowler.com/bliki/TestPyramid.html) approach
+to software testing, these should provide the majority of test coverage.
+
+Characteristics of unit tests:
+
+- they should be concerned with the logic within the class/module being tested,
+  and not test logic defined elsewhere;
+- they may mock dependent objects and/or assert that particular external
+  methods are called;
+- they should test all code paths through a class/module;
+- they should test that side effects intended by the code occur, for example writing
+  to the database or making API calls.
+
+## Feature tests
+
+Feature tests are used to assert that Content Publisher behaves correctly from
+the perspective of someone using the application. This ensures we're testing
+the system in a way representative of expected user behaviour. This makes the
+tests some of the most valuable of the application, however they are slower
+than other tests to run and can be difficult to debug. Therefore they are not
+intended to exhaustively test all the scenarios that can occur.
+
+Characteristics of feature tests:
+
+- they should test across multiple requests - consider a
+  [request test](#request-tests) to test a single request;
+- they should navigate the application using the same means as a user, for
+  example clicking links or submitting forms;
+- they shouldn't test dead ends in a user flow (such as validation or
+  permission issues); these are simpler tested in
+  [request tests](#request-tests);
+- incidental behaviour (for example flash messages) shouldn't be tested, this
```

Comment:
I've decided to remove this line as it seems to be an outlier that doesn't quite represent current/past patterns - it's somewhat borrowed from the Whitehall guidelines as an approach I agree with but is not in practice by us sufficiently.

To explain my perspective, I'd suggest that you want the feature test to assert that a user has completed their task in one way and not every visual side effect of the change - which would be under the definition of incidental. Hopefully when a user has performed a task there is an indication of their change on the outcome page that can be asserted against (e.g. a change in data). 

I guess as an example of something incidental I'd pick [content data link in publish test](https://github.com/alphagov/content-publisher/blob/master/spec/features/workflow/publish_spec.rb#L57-L61). It's something that happened because I published, but it doesn't only happen on publish so should every state test it, nor does it feel a place you'd expect to change a test if you were changing the content data admin link.

Reply:
Thanks for explaining. Leaving it out/flexible sounds good to me, although I like your example of the link to Content Data, as that's definitely not part of the "publish feature".